host.go

```diff
@@ -0,0 +1,205 @@
+package sparkplughost
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"log/slog"
+	"time"
+
+	"github.com/EvergenEnergy/sparkplughost/protobuf"
+	"github.com/eclipse/paho.mqtt.golang"
+	"google.golang.org/protobuf/proto"
+)
+
+type HostApplication struct {
+	mqttClient               mqtt.Client
+	hostID                   string
+	brokerURL                string
+	lastWillMessageTimestamp time.Time
+	logger                   *slog.Logger
+	edgeNodeManager          *edgeNodeManager
+	disconnectTimeout        time.Duration
+}
+
+func NewHostApplication(brokerURL, hostID string, opts ...Option) *HostApplication {
+	cfg := defaultConfig()
+	for _, opt := range opts {
+		opt(cfg)
+	}
+
+	return &HostApplication{
+		hostID:            hostID,
+		brokerURL:         brokerURL,
+		logger:            cfg.logger,
+		edgeNodeManager:   newEdgeNodeManager(cfg.metricHandler),
+		disconnectTimeout: cfg.disconnectTimeout,
+	}
+}
+
+// Run will connect to the mqtt broker and block until
+// ctx is canceled.
+func (h *HostApplication) Run(ctx context.Context) error {
+	if h.mqttClient != nil && h.mqttClient.IsConnected() {
+		h.mqttClient.Disconnect(uint(h.disconnectTimeout.Milliseconds()))
+	}
+
+	h.initClient()
+
+	if token := h.mqttClient.Connect(); token.Wait() && token.Error() != nil {
+		return token.Error()
+	}
+
+	// wait until the context is cancelled
+	<-ctx.Done()
+
+	// disconnect gracefully by publishing the death certificate before
+	// sending the disconnect command to the MQTT broker.
+	token := h.publishOnlineStatus(false, time.Now().UTC())
+	if token.Wait() && token.Error() != nil {
+		return token.Error()
+	}
+
+	h.mqttClient.Disconnect(uint(h.disconnectTimeout.Milliseconds()))
+	return nil
+}
+
+func (h *HostApplication) initClient() {
```

Comment:
Do you think we can make this configurable via yaml in the future?

Reply:
not sure I understand the question. You mean the mqtt options?

Reply:
The options we want to allow clients of this package to configure are already configurable via the `Options` passed in the constructor. The other ones (like using the clean session flag) we want to make sure they are set to their current values since they are needed to comply with the spec.

re. yaml, keep in mind that this will be a package that gets imported from other go repositories. If those want to read the config parameters (like the broker url or the host id) from a `yaml` file they should be able to, but this library shouldn't make that decision for them.